cfblog/models.py

```diff
@@ -0,0 +1,146 @@
+
+    def __unicode__(self):
+        return self.name
+
+
+STATUS_CHOICES = (
+    (1, _('Draft')),
+    (2, _('Public')),
+)
+
+
+class BaseCmsPage(models.Model):
+    template = models.ForeignKey(CmsPageTemplate, verbose_name=_('template'), null=True, blank=True)
+    auth_data = JSONField(verbose_name=_('author data'), default={}, blank=True)
+    public_data = JSONField(verbose_name=_('public data'), default={}, blank=True)
+    created_on = models.DateTimeField(_('created on'), auto_now_add=True)
+    modified_on = models.DateTimeField(_('modified on'), auto_now=True)
+    status = models.IntegerField(_('status'), choices=STATUS_CHOICES, default=1)
+    publish = models.DateTimeField(_('publish'), default=datetime.datetime.now)
+
+    class Meta:
+        abstract = True
+
+    @property
+    def is_public(self):
+        return self.status == 2
+
+    def publish_cms_content(self):
+        self.public_data = self.auth_data
+        self.status = 2
+        self.save()
+
+
+class CmsPage(BaseCmsPage):
+    url = models.CharField(_('url path'),
+                           max_length=255, db_index=True, unique=True, validators=[validate_url_path])
+
+    @cacheable('cms_page_{id}')
+    def html(self):
+        return get_public_data(self)
+
+    @stale_cache('cms_page_{id}')
+    def save(self, *args, **kwargs):
+        super(CmsPage, self).save(*args, **kwargs)
+
+    def clean(self):
+        from .views import cms_page_index
+        resolver_math = resolve(self.url)
+        if resolver_math.func is cms_page_index and not self.template:
+            raise ValidationError(_(
+                'Template cannot be empty for new urls'
+            ))
+
+        if resolver_math.func is not cms_page_index and self.template:
+            raise ValidationError(_(
+                'Template is not required for existing urls'
+            ))
+        super(CmsPage, self).clean()
+
+
+class CmsBlogCategory(models.Model):
+    """Category model."""
+    LEVEL_CHOICES = (
+        (1, 1),
+        (2, 2),
+    )
+    title = models.CharField(_('title'), max_length=100)
+    description = models.TextField(_('description'), blank=True)
+    slug = models.SlugField(_('slug'))
+    parent = models.ForeignKey('self', blank=True, null=True)
+    level = models.IntegerField(_('level'), choices=LEVEL_CHOICES)
+    is_active = models.BooleanField(_('is_active'), default=True)
+    objects = CmsBlogCategoryManager()
+
+    class Meta:
+        verbose_name_plural = _('cms blog page categories')
+        ordering = ('title',)
+
+    def __unicode__(self):
+        return self.title
+
+
+class CmsBlogPost(BaseCmsPage):
+    title = models.CharField(_('title'), max_length=200)
+    author = models.ForeignKey(settings.AUTH_USER_MODEL, blank=True, null=True)
+    category = models.ForeignKey('CmsBlogCategory')
+    tags = TagField()
+    slug = models.SlugField(_('slug'), max_length=70)
```

Comment:
This is not needed, lets put the whole relative URL here. Slug is like foo-bar. We want /foo-bar/baz/ here.